rust/src/dirdiff.rs

```diff
@@ -0,0 +1,273 @@
+    pub(crate) changed_dirs: FileSet,
+}
+
+impl Diff {
+    #[allow(dead_code)]
+    pub(crate) fn count(&self) -> usize {
+        self.added_files.len()
+            + self.added_dirs.len()
+            + self.removed_files.len()
+            + self.removed_dirs.len()
+            + self.changed_files.len()
+            + self.changed_dirs.len()
+    }
+
+    pub(crate) fn contains(&self, s: &str) -> bool {
+        self.added_files.contains(s)
+            || self.added_dirs.contains(s)
+            || self.removed_files.contains(s)
+            || self.removed_dirs.contains(s)
+            || self.changed_files.contains(s)
+            || self.changed_dirs.contains(s)
+    }
+}
+
+impl fmt::Display for Diff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "files(added:{} removed:{} changed:{}) dirs(added:{} removed:{} changed:{})",
+            self.added_files.len(),
+            self.removed_files.len(),
+            self.changed_files.len(),
+            self.added_dirs.len(),
+            self.removed_dirs.len(),
+            self.changed_dirs.len()
+        )
+    }
+}
+
+fn file_content_changed(
+    src: &openat::Dir,
+    dest: &openat::Dir,
+    path: &str,
+    expected_len: u64,
+) -> Result<bool> {
+    let mut remaining = expected_len;
+    let mut srcf = std::io::BufReader::new(src.open_file(path)?);
+    let mut destf = std::io::BufReader::new(dest.open_file(path)?);
+    let mut srcbuf = [0; 4096];
+    let mut destbuf = [0; 4096];
+    let bufsize = srcbuf.len();
+    while remaining > 0 {
+        let readlen = std::cmp::min(usize::try_from(remaining).unwrap_or(bufsize), bufsize);
+        let mut srcbuf = &mut srcbuf[0..readlen];
+        let mut destbuf = &mut destbuf[0..readlen];
+        srcf.read_exact(&mut srcbuf)?;
+        destf.read_exact(&mut destbuf)?;
+        if srcbuf != destbuf {
+            return Ok(true);
+        }
+        remaining -= readlen as u64;
+    }
+    Ok(false)
+}
+
+fn is_changed(
+    src: &openat::Dir,
+    dest: &openat::Dir,
+    path: &str,
+    srcmeta: &openat::Metadata,
+    destmeta: &openat::Metadata,
+) -> Result<bool> {
+    if srcmeta.permissions() != destmeta.permissions() {
+        return Ok(true);
+    }
+    Ok(match srcmeta.simple_type() {
+        openat::SimpleType::File => {
+            if srcmeta.len() != destmeta.len() {
+                true
+            } else {
+                file_content_changed(src, dest, path, srcmeta.len())?
+            }
+        }
+        openat::SimpleType::Symlink => src.read_link(path)? != dest.read_link(path)?,
+        openat::SimpleType::Other => false,
+        openat::SimpleType::Dir => false,
+    })
+}
+
+fn canonicalize_name<'a>(prefix: &str, name: &'a str) -> Cow<'a, str> {
+    if prefix.is_empty() {
+        Cow::Borrowed(name)
+    } else {
+        Cow::Owned(format!("{}/{}", prefix, name))
+    }
+}
+
+fn diff_recurse(
+    prefix: &str,
```

Comment:
Looking at the surrounding code, this looks like a `Option` (unless you were planning to bridge this to C++ in the future).